sklearn/decomposition/nmf.py

```diff
@@ -55,55 +58,125 @@ def check_non_negative(X, whom):
+    check_non_negative(A, whom)
+    if np.max(A) == 0:
+        raise ValueError('Array passed to %s is full of zeros.' % whom)
+
+
+def _safe_compute_error(X, W, H):
+    """Frobenius norm between X and W.H, safe for sparse array"""
+    if not sp.issparse(X):
+        error = norm(X - np.dot(W, H))
+    else:
+        sqnorm_X = np.dot(X.data, X.data)
+        norm_WHT = trace_dot(np.dot(np.dot(W.T, W), H), H)
+        cross_prod = trace_dot((X * H.T), W)
+        error = sqrt(sqnorm_X + norm_WHT - 2. * cross_prod)
+    return error
+
+
+def _check_string_param(sparseness, solver):
+    allowed_sparseness = (None, 'data', 'components')
+    if sparseness not in allowed_sparseness:
+        raise ValueError(
+            'Invalid sparseness parameter: got %r instead of one of %r' %
+            (sparseness, allowed_sparseness))
+
+    allowed_solver = ('auto', 'proj-grad', 'coordinate')
+    if solver not in allowed_solver:
+        raise ValueError(
+            'Invalid solver parameter: got %r instead of one of %r' %
+            (solver, allowed_solver))
+
+
+def _initialize_nmf(X, n_components, init=None, eps=1e-6,
                     random_state=None):
-    """NNDSVD algorithm for NMF initialization.
+    """Algorithms for NMF initialization.
 
-    Computes a good initial guess for the non-negative
-    rank k matrix approximation for X: X = WH
+    Computes an initial guess for the non-negative
+    rank k matrix approximation for X: X = W.H
 
     Parameters
     ----------
 
-    X : array, [n_samples, n_features]
+    X : array-like, shape (n_samples, n_features)
         The data matrix to be decomposed.
 
-    n_components : array, [n_components, n_features]
+    n_components : integer
         The number of components desired in the approximation.
 
-    variant : None | 'a' | 'ar'
-        The variant of the NNDSVD algorithm.
-        Accepts None, 'a', 'ar'
-        None: leaves the zero entries as zero
-        'a': Fills the zero entries with the average of X
-        'ar': Fills the zero entries with standard normal random variates.
-        Default: None
+    init :  None | 'random' | 'nndsvd' | 'nndsvda' | 'nndsvdar' | 'uniform'
+        Method used to initialize the procedure.
+        Default: 'nndsvdar' if n_components < n_features, otherwise random.
+        Valid options:
+
+            'nndsvd': Nonnegative Double Singular Value Decomposition (NNDSVD)
+                initialization (better for sparseness)
+            'nndsvda': NNDSVD with zeros filled with the average of X
+                (better when sparsity is not desired)
+            'nndsvdar': NNDSVD with zeros filled with small random values
+                (generally faster, less accurate alternative to NNDSVDa
+                for when sparsity is not desired)
+            'random': non-negative random matrices, scale with:
+                sqrt(X.mean() / n_components)
+            'uniform': matrices filled with the same value:
```

Comment:
'uniform' is initialization by the same value everywhere, but I removed it in this PR.
I will do a specific PR to address the initialization schemes and run some benchmarks.
And I will also look at the random subset of rows/cols.